android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/ext/PopupWindowTest.kt

```diff
@@ -126,35 +137,200 @@ class PopupWindowTest {
-        verify(popupWindow).showAsDropDown(anchor, 0, 0)
+        anchor = createMockViewWith(0, 0)
+
+        popupWindow.displayPopup(menuContentView, anchor, Orientation.UP)
+
+        // At location
+        verify(popupWindow, never()).showAtLocation(any(), anyInt(), anyInt(), anyInt())
+        // As DOWN menu
+        verify(popupWindow, never()).showAsDropDown(any())
+        // As UP menu
+        verify(popupWindow, never()).showAsDropDown(any(), anyInt(), anyInt())
+    }
+
+    @Test
+    fun `WHEN up menu container height fits max app height THEN show popup up with original height`() {
+        val mockRecyclerView = createMockRecyclerViewWith(1)
+        doReturn(mockRecyclerView).`when`(menuContentView)
+            .findViewById<RecyclerView>(R.id.mozac_browser_menu_recyclerView)
+
+        val x = 0
+        anchor = createMockViewWith(x, 130)
+        val containerHeight = mockRecyclerView.measuredHeight.plus(1)
+        doReturn(containerHeight).`when`(menuContentView).measuredHeight
+
+        popupWindow.displayPopup(menuContentView, anchor, Orientation.UP)
+
+        verify(popupWindow).height = containerHeight
+        verify(popupWindow).showAsDropDown(anchor, x, -containerHeight)
+    }
+
+    @Test
+    fun `WHEN up menu container height doesn't fit within max app height & has less than required height THEN show popup up with calculated height`() {
+        val mockRecyclerView = createMockRecyclerViewWith(itemCount = 7, recyclerViewHeight = 50)
+        doReturn(mockRecyclerView).`when`(menuContentView)
+            .findViewById<RecyclerView>(R.id.mozac_browser_menu_recyclerView)
+
+        val x = 0
+        anchor = createMockViewWith(x, 40)
+        val containerHeight = mockRecyclerView.measuredHeight.plus(1)
+        doReturn(containerHeight).`when`(menuContentView).measuredHeight
+
+        popupWindow.displayPopup(menuContentView, anchor, Orientation.UP)
+
+        // The configuration will force the 'number of items that fit exactly with overflow' to
+        // use the minimum number of menu items.
+        val expectedMenuContainerHeight = 39
+        verify(popupWindow).height = expectedMenuContainerHeight
+        verify(popupWindow).showAsDropDown(anchor, x, -expectedMenuContainerHeight)
+    }
+
+    @Test
+    fun `WHEN up menu container height doesn't fit within max app height & at the required height THEN show popup up with calculated height`() {
+        val mockRecyclerView = createMockRecyclerViewWith(8, recyclerViewHeight = 50)
+        doReturn(mockRecyclerView).`when`(menuContentView)
+            .findViewById<RecyclerView>(R.id.mozac_browser_menu_recyclerView)
+
+        val x = 0
+        anchor = createMockViewWith(x, 40)
+        val containerHeight = mockRecyclerView.measuredHeight.plus(1)
+        doReturn(containerHeight).`when`(menuContentView).measuredHeight
+
+        popupWindow.displayPopup(menuContentView, anchor, Orientation.UP)
+
```

Comment:
magic number with no comments :) 
50 / 8 = 6 – item height
40 anchor height, so 6 items could fit. we crop the last one, so it's 33, plus 1 padding == 34!